Model: Granite MoE shared

gabe-l-hart opened the pull request:
## Description

This PR adds support for the `GraniteMoEShared` architecture, matching the implementation in [transformers](https://github.com/huggingface/transformers/tree/main/src/transformers/models/granitemoeshared). The model is an iteration on top of `GraniteMoE` and adds a shared expert to each MoE layer.

**NOTE**: There is not a public model with this architecture for testing yet, but it is a key building block for the just-released [Granite 4](https://huggingface.co/ibm-granite/granite-4.0-tiny-preview) architecture.

gabe-l-hart committed:
feat: Add GGUF conversion for granitemoeshared

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: hparam and arch plumbing for granitemoeshared

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Split MoE fused tensors for shared experts in conversion

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: First WIP cut at model arch in cpp

The hparam and architecture plumbing should be correct, but the
implementation of the shared experts seems to still be broken.

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Cleaner (maybe more correct?) splitting for gate/up

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix the input to the shared experts

I had misread that the shared experts take the inputs _before_ the standard
MoE layer and was feeding the output of the MoE to the shared experts.

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ngxson commented on src/llama-model.cpp:
```diff
@@ -1753,6 +1760,13 @@ bool llama_model::load_tensors(llama_model_loader & ml) {
+
+                            // For Granite MoE Shared
+                            if (arch == LLM_ARCH_GRANITE_MOE_SHARED) {
+                                layer.ffn_gate_shexp = create_tensor(tn(LLM_TENSOR_FFN_GATE_SHEXP, "weight", i), {n_embd, hparams.n_ff_shexp}, 0);
```
I think you can simply check `hparams.n_ff_shexp > 0` then load these tensors. The value of `hparams.n_ff_shexp` will be 0 if it's not written in gguf

This way, you can remove 2/3 code of this PR, making it much easier

gabe-l-hart replied:
Ah, yes, this would be _much_ simpler! All of the granite architectures are highly related and slight tweaks on other architectures. In `transformers`, the approach has always been to keep the architectures isolated and use their `modular` approach for shared code. I know that policy is less strict here, but wasn't sure how much to lean into collapsing all of these architectures into a single architecture string versus keeping them separate and sharing code underneath. Is there a strong preference here going forward?

gabe-l-hart replied:
Makes sense. My personal preference is to keep a 1:1 mapping between architecture names in `huggingface`, then have maximal code reuse on the backend. That is a very weakly held preference, though, so I'm happy to take whatever direction is best from the maintainers' POV.

ngxson replied:
@gabe-l-hart Not sure if you have a near deadline to merge this PR (?)

We never aim for 1:1 mapping between HF `model_type` <> llama.cpp arch name from the beginning, so it's probably fine to consider this new arch as a variant of `GRANITE_MOE` with `n_ff_shexp > 0`. I'm not sure in which case the 1:1 mapping will be useful?

In anw, after second thought, I have no strong opinion whether this need to be a dedicated arch name or not, you can probably keep it this way if you like. But maybe we should move the cgraph of granite family to a dedicated `build_*` function. My main concern is that we're adding things on top of `build_llama` which make it not very easy to read

gabe-l-hart committed:
fix: Avoid architecture-specific checks for Granite MoE Shared

This is a cleaner way that will allow more flexibility in architecture
strings going forward.

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: Split granite architectures out of llm_build_llama

This helps de-clutter the llama-family graph construction and allows
granite to diverge further (in preparation for Granite 4).

NOTE: I removed the granite scale factors from llm_build_deci because they
appear to only be there as copy-paste from llm_build_llama. The HF config
does not seem to set those values:
https://huggingface.co/Deci/DeciLM-7B/blob/main/config.json

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix compiler warning about uninitialized inp_pos

This should not have been reachable, but it warns on some compliers

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov commented:
I agree with @ngxson that adding separate arch for MoE-shared is a bit redundant, but it's OK either way. The separate build function for Granite models is good refactoring.

I think we are ready to merge. @gabe-l-hart Are the models out?

gabe-l-hart committed:
fix: Consoladate GraniteMoEShared into GraniteMoE for conversion

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Consolidate GraniteMoEShared into GraniteMoE on the c++ side

Branch: GraniteMoEShared

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov commented on src/llama-model.cpp:
```diff
@@ -12210,6 +12193,195 @@ struct llm_build_arwkv7 : public llm_build_rwkv7_base {
+
+                if (use_rope) {
+
+                    if (!inp_pos) {
+                        inp_pos = build_inp_pos();
+                    }
```
@gabe-l-hart This should be moved at the beginning of the loop function, before the loop over the layers, to avoid creating the tensor for each layer.

gabe-l-hart replied:
Thanks for pointing this out. My thought with putting it there was that it would be lazily initialized on the first layer if `use_rope` is `true` and then not re-initialized on subsequent loop iterations because of the `if` check. Is there something I'm missing about how this tensor is used below that would cause the pointer to be `nullptr` on later loop iterations?

Regardless, there's no harm in putting it back at the top with the `use_rope` check, so I can get a small PR in place to do that or just lump the change in with my Granite 4 branch.

ggerganov replied:
I got somehow confused - it won't create the tensor. Still, it's better to move it before the loop for consistency with the other build functions.

gabe-l-hart replied:
Makes sense! Quick PR to fix it: https://github.com/ggml-org/llama.cpp/pull/13538
